kube-runtime/src/utils/delayed_init.rs

```diff
@@ -0,0 +1,221 @@
+        let _ = self.0.send(value);
+    }
+}
+impl<T> Debug for Initializer<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("delayed_init::Initializer")
+            .finish_non_exhaustive()
+    }
+}
+
+/// A value that must be initialized by an external writer
+///
+/// Can be considered equivalent to a [`channel::oneshot`] channel, except for that
+/// the value produced is retained for subsequent calls to [`Self::get`].
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct DelayedInit<T> {
+    state: RwLock<ReceiverState<T>>,
+    // A test-only hook to let us create artificial race conditions
+    #[cfg(test)]
+    #[allow(clippy::type_complexity)]
+    #[derivative(Debug = "ignore")]
+    test_hook_start_of_slow_path: Box<dyn Fn(&mut RwLockWriteGuard<ReceiverState<T>>) + Send + Sync>,
+}
+#[derive(Debug)]
+enum ReceiverState<T> {
+    Waiting(channel::oneshot::Receiver<T>),
+    Ready(Result<T, InitDropped>),
+}
+impl<T> DelayedInit<T> {
+    /// Returns an empty `DelayedInit` that has no value, along with a linked [`Initializer`]
+    #[must_use]
+    pub fn new() -> (Initializer<T>, Self) {
+        let (tx, rx) = channel::oneshot::channel();
+        (Initializer(tx), DelayedInit {
+            state: RwLock::new(ReceiverState::Waiting(rx)),
+            #[cfg(test)]
+            test_hook_start_of_slow_path: Box::new(|_| ()),
+        })
+    }
+}
+impl<T: Clone + Send + Sync> DelayedInit<T> {
+    /// Wait for the value to be available and then return it
+    ///
+    /// Calling `get` again if a value has already been returned is guaranteed to return (a clone of)
+    /// the same value.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the associated [`Initializer`] has been dropped before calling [`Initializer::init`].
+    pub async fn get(&self) -> Result<T, InitDropped> {
+        Get(self).await
+    }
+}
+
+// Using a manually implemented future because we don't want to hold the lock across poll calls
+// since that would mean that an unpolled writer would stall all other tasks from being able to poll it
+struct Get<'a, T>(&'a DelayedInit<T>);
+impl<'a, T> Future for Get<'a, T>
+where
+    T: Clone,
+{
+    type Output = Result<T, InitDropped>;
+
+    #[tracing::instrument(name = "DelayedInit::get", level = "trace", skip(self, cx))]
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let read_lock = self.0.state.read().unwrap();
+        if let ReceiverState::Ready(v) = &*read_lock {
+            trace!("using fast path, value is already ready");
+            Poll::Ready(v.clone())
+        } else {
+            trace!("using slow path, need to wait for the channel");
+            // IMPORTANT: Make sure that the optimistic read lock has been released already
+            drop(read_lock);
+            let mut state = self.0.state.write().unwrap();
```

Comment:
So afaikt the race possibility is basically between the drop and reclaiming (some other poll could get interleaved and also end up in ready to write).

Is the hit of just using a straight mutex on the state, and then matching on that too big? There's a lot of complexity here (particularly on the test side, which leaks into the main type).

Reply:
Yes, that's correct.

Reply:
There's no more optimistic read lock to worry about anyway.